### PR TITLE
Add migrate command

### PR DESCRIPTION
**Pull request by mja00**
This command will allow users to update their packs to newer loader and Minecraft versions. When updating Minecraft it will attempt to also update any mods that it can.

### CODE REVIEW

**Comment by mja00** [comment]
Just a note to self, gotta figure out a way to get the recommended version for Forge. Long long ago they had an API for their downloads page but I don't know if it still exists.

**Comment by mja00** [comment]
I believe this is ready for review. I had to move some stuff into a shared package as well as make the Update command publicly accessible to the migrate package that way we weren't rewriting the same code over again.

File: core/versionutil.go [hunk 1]
```diff
@@ -96,7 +96,7 @@ func FetchMavenVersionPrefixedList(url string, friendlyName string) func(mcVersi
 		}
 		// Sort list to get largest version
 		flexver.VersionSlice(out.Versioning.Versions.Version).Sort()
-		return allowedVersions, allowedVersions[len(allowedVersions)-1], nil
+		return allowedVersions, allowedVersions[0], nil
```
Comment (comp500) [inline core/versionutil.go:99]: ~~What's the reason for this change? The second return value should be the newest version (i.e. the last version when sorted)~~ Discussed on discord - the previous line needs to be fixed, whether you want to do so in this PR or push it separately.

File: migrate/loader.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	"github.com/packwiz/packwiz/core"
+	"github.com/spf13/cobra"
+	"golang.org/x/exp/slices"
+	"os"
+	"strings"
+)
+
+var loaderCommand = &cobra.Command{
+	Use:   "loader",
+	Short: "Migrate your loader versions to newer versions.",
```
Comment (comp500) [inline migrate/loader.go:14]: Could you specify "modloader"? Might make more sense to make this singular rather than plural as well, as the code ensures you only have 1 loader.

File: migrate/loader.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+			fmt.Printf("Error loading pack: %s\n", err)
+			os.Exit(1)
+		}
+		// Get our current loader
+		currentLoader := modpack.GetLoaders()
+		// Do some sanity checks on the current loader slice
+		if len(currentLoader) == 0 {
+			fmt.Println("No loader is currently set in your pack.toml!")
+			os.Exit(1)
+		} else if !slices.Contains(currentLoader, "quilt") && len(currentLoader) > 1 {
```
Comment (comp500) [inline migrate/loader.go:33]: Edge case: if Quilt is being used, the check for more than 1 loader is ignored (so the user could have 3 loaders?)

File: migrate/loader.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			// Check if it's a no such file or directory error
+			if os.IsNotExist(err) {
+				fmt.Println("No pack.toml file found, run 'packwiz init' to create one!")
+				os.Exit(1)
+			}
+			fmt.Printf("Error loading pack: %s\n", err)
+			os.Exit(1)
+		}
+		// Get our current loader
+		currentLoader := modpack.GetLoaders()
```
Comment (comp500) [inline migrate/loader.go:28]: Could you make this variable name plural to indicate that it is a slice.. or make another variable with `currentLoaders[0]` after the sanity check?

File: migrate/loader.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		// Do some sanity checks on the current loader slice
+		if len(currentLoader) == 0 {
+			fmt.Println("No loader is currently set in your pack.toml!")
+			os.Exit(1)
+		} else if !slices.Contains(currentLoader, "quilt") && len(currentLoader) > 1 {
+			fmt.Println("You have multiple loaders set in your pack.toml, this is not supported!")
+			os.Exit(1)
+		} else if slices.Contains(currentLoader, "quilt") {
+			// We have quilt, so we need to remove fabric from the loaders list
```
Comment (comp500) [inline migrate/loader.go:37]: Hmm... I might make a version of `GetLoaders` that doesn't add Fabric so you don't have to do this. Alternatively you could work directly on the `modpack.Versions` map...

File: migrate/loader.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	"github.com/spf13/cobra"
+	"golang.org/x/exp/slices"
+	"os"
+	"strings"
+)
+
+var loaderCommand = &cobra.Command{
+	Use:   "loader",
+	Short: "Migrate your loader versions to newer versions.",
+	Args:  cobra.ExactArgs(1),
```
Comment (comp500) [inline migrate/loader.go:15]: Please document the arguments to this command (even just something in the `Use` string would be much better)

File: migrate/loader.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+			fmt.Println("Currently updating to the recommended loader version is not supported!")
+			os.Exit(1)
+		} else {
+			fmt.Println("Updating to explicit loader version")
+			// This one is easy :D
+			versions, _, loader := getVersionsForLoader(currentLoader[0], mcVersion)
+			// Check if the loader happens to be Forge, since there's two version formats
+			if loader.Name == "forge" {
+				var wantedVersion string
+				// Check if we have a "-" in the version
```
Comment (comp500) [inline migrate/loader.go:73]: I think the version list getter already strips versions for you? The init command doesn't seem to do any post-processing anyway.

File: migrate/minecraft.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+		// Prompt the user if they want to update the loader too while they're at it.
+		if cmdshared.PromptYesNo("Would you like to update your loader version to the latest version for this Minecraft version? [Y/n] ") {
+			// We'll run the loader command to update to latest
+			loaderCommand.Run(loaderCommand, []string{"latest"})
+		}
+		// Prompt the user to update their mods too.
+		if cmdshared.PromptYesNo("Would you like to update your mods to the latest versions for this Minecraft version? [Y/n] ") {
+			// Run the update command
+			viper.Set("update.all", true)
+			packCmd.UpdateCmd.Run(packCmd.UpdateCmd, []string{})
```
Comment (comp500) [inline migrate/minecraft.go:63]: Hmm.... I don't like this, but it works I guess. I would prefer splitting out shared functionality into more `cmdshared` functions, particularly as it makes dependency cycles less likely to occur.

**Review by comp500 — CHANGES_REQUESTED** [review]
Thanks for the PR! I have a few suggestions :)

File: migrate/loader.go (hunk 7 above)
Comment (mja00) [reply migrate/loader.go:96]: This is input from the user, which'll let them either do `1.19.4-44.1.23` or `44.1.23` for example, since Forge uses both formats interchangeable everywhere.

File: migrate/minecraft.go (hunk 8 above)
Comment (mja00) [reply migrate/minecraft.go:63]: I didn't like it either and yeah splitting it out into  `cmdshared` functions would definitely be better. I assume, above where I run `loaderCommand.Run(loaderCommand, []string{"latest"})` is fine since it's in the same package?

File: migrate/loader.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+			if key == "minecraft" {
+				continue
+			}
+			currentLoaders = append(currentLoaders, key)
+		}
+		// Do some sanity checks on the current loader slice
+		if len(currentLoaders) == 0 {
+			fmt.Println("No loader is currently set in your pack.toml!")
+			os.Exit(1)
+		} else if (!slices.Contains(currentLoaders, "quilt") && len(currentLoaders) > 1) || (slices.Contains(currentLoaders, "quilt") && len(currentLoaders) > 2) {
```
Comment (comp500) [inline migrate/loader.go:42]: I don't think this needs to special-case quilt anymore :)

File: migrate/loader.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+				fmt.Println("No pack.toml file found, run 'packwiz init' to create one!")
+				os.Exit(1)
+			}
+			fmt.Printf("Error loading pack: %s\n", err)
+			os.Exit(1)
+		}
+		// Get our current loader, would use the modpack.GetLoaders function
+		// but, it supplements Quilt with Fabric, which isn't needed for this
+		var currentLoaders []string
+		// Add all the keys from modpack.Versions
```
Comment (comp500) [inline migrate/loader.go:30]: Probably a good idea to filter this by the supported loaders in `core.ModLoaders`; old versions of packwiz [used a yarn field](https://github.com/Fibercraft/Temporary-Modpack/blob/f33038291c48e7978b47b2a4453f11b768fae9e4/pack.toml#L11) and unsup adds [it's own version field for format extensions](https://git.sleeping.town/unascribed-tools/unsup/wiki/Packwiz-extensions).

File: migrate/loader.go (hunk 7 above)
Comment (comp500) [reply migrate/loader.go:96]: Fair... this could probably be extracted into a `cmdshared` function and used in the init command too

File: migrate/minecraft.go (hunk 8 above)
Comment (comp500) [reply migrate/minecraft.go:63]: Yeah that's not as bad. Don't bother splitting the update command out into `cmdshared` for now, it's probably something which deserves a separate rewrite anyway.

**Review by comp500 — CHANGES_REQUESTED** [review]

File: migrate/loader.go (hunk 9 above)
Comment (mja00) [reply migrate/loader.go:42]: Ah yeah you right lol

File: migrate/loader.go (hunk 7 above)
Comment (mja00) [reply migrate/loader.go]: Abstracted out into `cmdshared.GetRawForgeVersion`.

File: migrate/minecraft.go (hunk 8 above)
Comment (mja00) [reply migrate/minecraft.go:63]: Fair, I'll leave it as is for now then.

**Review by comp500 — APPROVED** [review]
Looks good to me! The original issue https://github.com/packwiz/packwiz/issues/40 suggested telling users which mods are incompatible with the new version - I'll make a new issue for that feature specifically.
